File: solution/simulation/20058/main.cpp
// Authored by : tony9402
// Co-authored by : -
// Link : http://boj.kr/4e60b2336bfa4e3192b9aef4eb5a3f6b
#include<bits/stdc++.h>

using namespace std;

const int dy[] = {-1,1,0,0};
const int dx[] = {0,0,-1,1};

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);
    
    int K, Q; cin >> K >> Q;
    int N = 1 << K;
    vector<vector<int>> Map(N, vector<int>(N));
    for(int i=0;i<N;i++){
        for(int j=0;j<N;j++){
            cin >> Map[i][j];
        }
    }

    while(Q--) {
        int L; cin >> L;
        int M = 1 << L;
        vector<vector<int>> tmp = Map;
        for(int i=0;i<N;i+=M){
            for(int j=0;j<N;j+=M){
                for(int m=0;m<M;m++){
                    for(int n=0;n<M;n++){
                        Map[i+n][j+M-1-m] = tmp[i+m][j+n];
                    }
                }
            }
        }
        tmp = Map;
        for(int i=0;i<N;i++){
            for(int j=0;j<N;j++){
                if(Map[i][j] == 0)continue;
                int cnt = 0;
                for(int k=0;k<4;k++) {
                    int qy = i + dy[k];
                    int qx = j + dx[k];
                    if(0 > qy || qy >= N || 0 > qx || qx >= N) continue;
                    if(Map[qy][qx] != 0) cnt ++;
                }
                if(cnt < 3) tmp[i][j] --;
            }
        }
        Map = tmp;
    }

    int total = 0;
    int mx    = 0;
    for(int i=0;i<N;i++){
        for(int j=0;j<N;j++){
            if(Map[i][j] == 0)continue;
            total += Map[i][j];
            int cnt = 0;
            Map[i][j] = 0;
            queue<pair<int, int>> q;
            q.emplace(i, j);
            while(!q.empty()) {
                auto [y, x] = q.front(); q.pop();
                cnt++;
                for(int k=0;k<4;k++){
                    int qy = y + dy[k];
                    int qx = x + dx[k];
                    if(0 > qy || qy >= N || 0 > qx || qx >= N) continue;
                    if(Map[qy][qx] == 0)continue;
                    q.emplace(qy, qx);
                    total += Map[qy][qx];
                    Map[qy][qx] = 0;
                }
            }
            mx = max(mx, cnt);
        }
    }
    
    cout << total << '\n';
    cout << mx;

    return 0;
}

File: solution/simulation/20058/main.py
# Authored by : tony9402
# Co-authored by : -
# Link : http://boj.kr/f0c88b8e08204d33a7e63a384b5fdc4c
import sys
from copy import deepcopy
from collections import deque

def input():
    return sys.stdin.readline().rstrip()

dy = (-1, 1, 0, 0)
dx = ( 0, 0,-1, 1)
K, Q = map(int, input().split())
N = 1 << K

arr = list()
for i in range(N):
    arr.append(list(map(int, input().split())))

CMD = list(map(int, input().split()))

for _L in CMD:
    L = 1 << _L
    tmp = deepcopy(arr)
    for i in range(0, N, L):
        for j in range(0, N, L):
            for n in range(0, L):
                for m in range(0, L):
                    arr[i + m][j + L - 1 - n] = tmp[i + n][j + m]

    tmp = deepcopy(arr)
    for i in range(N):
        for j in range(N):
            if arr[i][j] == 0:
                continue
            cnt = 0
            for k in range(4):
                y = i + dy[k]
                x = j + dx[k]
                if 0 > y or y >= N or 0 > x or x >= N:
                    continue
                if arr[y][x] != 0:
                    cnt += 1

            if cnt < 3:
                tmp[i][j] -= 1
    arr = deepcopy(tmp)

total = 0
mx    = 0

for i in range(N):
    for j in range(N):
        if arr[i][j] == 0:
            continue
        cnt = 0
        que = deque()
        que.append((i, j))
        total += arr[i][j]
        arr[i][j] = 0
        while len(que) != 0:
            cy, cx = que.popleft()
            cnt += 1
            for k in range(4):
                y = cy + dy[k]
                x = cx + dx[k]
                if 0 > y or y >= N or 0 > x or x >= N or arr[y][x] == 0:
                    continue
                que.append((y, x))
                total += arr[y][x]
                arr[y][x] = 0
        mx = max(mx, cnt)

print(f"{total}\n{mx}")
